src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -8333,6 +8379,312 @@ private StatementSyntax ParseLocalDeclarationStatement()
+
+            try
+            {
+                var openParen = this.EatToken(SyntaxKind.OpenParenToken);
+
+                if (this.CurrentToken.Kind != SyntaxKind.CloseParenToken)
+                {
+                    var variable = ParseDeconstructionDeclaration(withEquals: false, topLevel: false);
+
+                    if ((object)variable != null)
+                    {
+                        list.Add(variable);
+                    }
+
+                    while (this.CurrentToken.Kind == SyntaxKind.CommaToken && (object)variable != null)
+                    {
+                        var comma = this.EatToken(SyntaxKind.CommaToken);
+                        list.AddSeparator(comma);
+
+                        variable = ParseDeconstructionDeclaration(withEquals: false, topLevel: false);
+
+                        if ((object)variable != null)
+                        {
+                            list.Add(variable);
+                        }
+                    }
+                }
+
+                var closeParen = this.EatToken(SyntaxKind.CloseParenToken);
+
+                SyntaxToken equals = null;
+                ExpressionSyntax expression = null;
+                if (withEquals)
+                {
+                    equals = this.EatToken(SyntaxKind.EqualsToken);
+                    if (equals.IsMissing)
+                    {
+                        return null;
+                    }
+
+                    expression = this.ParseExpressionCore();
+                }
+
+                var deconstruction = _syntaxFactory.VariableDeconstructionDeclarator(openParen, list, closeParen, equals, expression);
+                var result = _syntaxFactory.VariableDeclaration(deconstruction);
+
+                if (!result.ContainsDiagnostics && list.Count < 2)
+                {
+                    result = this.AddError(result, ErrorCode.ERR_DeconstructTooFewElements);
+                }
+
+                return result;
+            }
+            finally
+            {
+                _pool.Free(list);
+            }
+        }
+
+        /// <summary>
+        /// Parses the var form of deconstruction. For instance, `var (x, y) = ...`
+        /// </summary>
+        private VariableDeclarationSyntax ParseDeconstructionVarForm(bool withEquals)
+        {
+            Debug.Assert(this.CurrentToken.IsVar);
+
+            var varType = ParseType(parentIsParameter: false);
+
+            var identifiers = ParseDeconstructionIdentifiers(withEquals);
+
+            return _syntaxFactory.VariableDeclaration(SyntaxKind.DeconstructionDeclaration, varType, default(SeparatedSyntaxList<VariableDeclaratorSyntax>), identifiers);
+        }
+
+        /// <summary>
+        /// Parses `(..., ..., ...)` where each part is either an identifier, or nested deconstruction identifiers
+        /// </summary>
+        private VariableDeconstructionDeclaratorSyntax ParseDeconstructionIdentifiers(bool withEquals)
```

Comment:
It looks `ParseDeconstructionListForm` and `ParseDeconstructionIdentifiers` are similar except for the call to parse each declaration or identifier. Is it worth creating a single method with a flag?